### ISSUE TITLE
Home IDP Discovery not working as expected

### ISSUE THREAD

**Issue by ankur-tj** [p2-inc/keycloak-orgs#258]
Hello,

I have two demo instances of Keycloak Org. The first is the master instance and the 2nd is the IDP. I configured the IDP in my master instance and linked it to an organization. 
I have my Angular app registered as a client in my master keycloak and the authentication flow used is Org browser flow. I added the extra step of Home discovery but it doesn't work as expected. I tried deleting few steps and also tried inserting the IDP step at random serial order so I don't have exact steps to reproduce the issue. After trying a few combinations of placing the IDP step, I was finally redirected to the IDP for the 1st time and I was able to successfully get the token and got redirected to my Angular app but when I logged out and tried to login again, this time it directly sent me to my angular app instead of redirecting to IDP (no validation at all)

Basically, I need IDP discovery as 1st step. If there is no matching org found then show the password field. I also need the OTP step.

1. Can someone please tell me which authentication flow handled this or help me create a new one with exact steps?
2. Can someone also tell me how to block the user info getting copied to my master Keycloak when someone logs in through IDP?
3. I also want to understand the multiple organization flow. I added the same user to multiple orgs but I was not prompted with the choice of Organization at the time of login.

Please reply back as we are in the middle of a decision making process of whether to go with the Keycloak org hosting plan or OKTA

**Comment by xgp** [comment]
1. See the following video for a simple setup 

https://github.com/p2-inc/keycloak-orgs/assets/244253/1d9da9d5-f4db-4a1d-8984-d38462a9e2be


2. Identity brokering in Keycloak creates a local user. Find more information about everything that happens in the [First Login Flow](https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker_first_login) portion of the Keycloak documentation. 
3. For more information about using a mechanism of choosing from multiple organizations, see the documentation on the [Active Organization](https://github.com/p2-inc/keycloak-orgs/blob/main/docs/active-organization.md) features.

> when I logged out and tried to login again, this time it directly sent me to my angular app instead of redirecting to IDP (no validation at all)

This means you were still logged in to the remote IdP with an active SSO session. This worked as expected, because you are not supposed to have to log in if you have an active SSO session. If you want to have a logout from your app propagate the remote IdP, make sure you have the following settings checked in your IdP configuration:
![image](https://github.com/p2-inc/keycloak-orgs/assets/244253/0c62b83d-a2f8-46a6-a513-bb5f61a90f50)

> Please reply back as we are in the middle of a decision making process of whether to go with the Keycloak org hosting plan or OKTA

Good luck with your evaluation of Keycloak.

**Comment by ankur-tj** [comment]
First of all thanks for answering my queries. 

I have gone through the video and I tried doing the same setup as yours but I could find the step/condition called 'Organization Identity-First login'.

This is how it looks like.
![image](https://github.com/p2-inc/keycloak-orgs/assets/22630809/8ca7430b-7697-488e-8fd6-c89c0a167c61)
![image](https://github.com/p2-inc/keycloak-orgs/assets/22630809/6a09c92f-c28d-456d-ae0f-050e5527770f)

Now the issue that I am facing is that I am able to redirect the user to Organization configured IDP for the first time but once the user logs out and tries to log in again, the below endpoint returns the error.

{{url}}/realms/microservices/login-actions/authenticate?session_code=v8-Dxncx8xS4OaumBigHhP2PWCufzeRgAd3_NIceGx0&execution=315da79d-7b37-4ae1-b658-94bb10a65488&client_id=angularapp&tab_id=Uvww_94tL3o&client_data=eyJydSI6Imh0dHA6Ly9sb2NhbGhvc3Q6NDIwMCIsInJ0IjoiY29kZSIsInN0IjoiU1RKRWJHdGFWak5SYW1WYU5Hb3pRek13VVZGSFZtTm1kSEpOV25CNlYzcFJhM2s1ZUdWUVNXcFZXVkUyIn0

![image](https://github.com/p2-inc/keycloak-orgs/assets/22630809/8e7c7d04-c274-4695-b704-93ea205416dd)

**Comment by xgp** [comment]
> 'Organization Identity-First login'

That stuff's all irrelevant. Ignore it.

Here's a real simplified version that works:
- flow
![image](https://github.com/p2-inc/keycloak-orgs/assets/244253/a5be3bd3-e4b9-4353-b412-18dc6cb1b535)
- home idp config
![image](https://github.com/p2-inc/keycloak-orgs/assets/244253/9f58caf5-663a-44e3-afd7-9473a1df431c)

Regarding the second login problem, that's either a problem with your Identity Provider configuration or something that comes up when OTP isn't properly configured. That's not specific to this extension. Please consult the [Keycloak documentation](https://www.keycloak.org/docs/latest/server_admin/index.html) or [repository](https://github.com/keycloak/keycloak/issues)/[forum](https://keycloak.discourse.group/) for help there.

> username field was auto-filled

That's how it looks with Keycloak when you pass the `login_hint`. How it behaves when you pass that parameter will be up to the identity provider. But, if you are seeing the username in your Keycloak example, it is working.

**Comment by ankur-tj** [comment]
Thanks for sharing this really simplified version where everything is working as expected. 

Is there a configuration in the Home IDP Discovery step to redirect the user to their IDP only if there exists a user with the same username in Keycloak?
Or if there is an alternative way to check this, then please share that also.

**Comment by xgp** [comment]
No

**Comment by ankur-tj** [comment]
Actually, this is possible, you just have to set this flag to true.

![image](https://github.com/user-attachments/assets/034fd795-eb6f-4ecf-88a3-55433a88784e)

The only issue here is that the error message is not user friendly. 

![image](https://github.com/user-attachments/assets/34fe2e1e-d730-48ce-9b83-4589f46b2876)

If there is a way to customize this message, please let me know.

**Comment by xgp** [comment]
messages can be updated in custom themes https://www.keycloak.org/docs/latest/server_development/index.html#_themes

**Comment by ankur-tj** [comment]
Thanks for sharing the link. I tried overwriting the message `identityProviderUnexpectedErrorMessage` through realm attributes `_providerConfig.theme.login.messages.identityProviderUnexpectedErrorMessage` but it's not reflecting. I have selected attributes theme for login page. I believe the issue is with the path to the messageKey through providerConfig. 

Could you please throw some light on this?

**Comment by xgp** [comment]
That is an experimental part of the `keycloak-themes` extension. I would recommend building a custom theme instead.

**Comment by ankur-tj** [comment]
I have 3 more attributes and those are working fine. The problem is only with overwriting the message. Is there no other way apart from building a custom theme?

![image](https://github.com/user-attachments/assets/fc4e4f30-169f-4f45-b89c-71ac4bcd7c72)

**Comment by xgp** [comment]
As I said, it's **experimental**. I don't have an idea why some keys work and some don't. If you'd like to help us figure it out, please look into it and submit a PR.
